sphinx/factory.py

```diff
@@ -0,0 +1,228 @@
+# -*- coding: utf-8 -*-
+"""
+    sphinx.factory
+    ~~~~~~~~~~~~~~
+
+    Sphinx component factory.
+
+    Gracefully adapted from the TextPress system by Armin.
+
+    :copyright: Copyright 2007-2016 by the Sphinx team, see AUTHORS.
+    :license: BSD, see LICENSE for details.
+"""
+from __future__ import print_function
+
+import traceback
+
+from pkg_resources import iter_entry_points
+from six import itervalues
+
+from sphinx.errors import ExtensionError, SphinxError, VersionRequirementError
+from sphinx.extension import Extension
+from sphinx.domains import ObjType
+from sphinx.domains.std import GenericObject, Target
+from sphinx.locale import _
+from sphinx.roles import XRefRole
+from sphinx.util import logging
+from sphinx.util.docutils import directive_helper
+
+if False:
+    # For type annotation
+    from typing import Any, Callable, Dict, Iterator, List, Type  # NOQA
+    from docutils import nodes  # NOQA
+    from docutils.parsers import Parser  # NOQA
+    from sphinx.application import Sphinx  # NOQA
+    from sphinx.builders import Builder  # NOQA
+    from sphinx.domains import Domain, Index  # NOQA
+    from sphinx.environment import BuildEnvironment  # NOQA
+
+logger = logging.getLogger(__name__)
+
+# list of deprecated extensions. Keys are extension name.
+# Values are Sphinx version that merge the extension.
+EXTENSION_BLACKLIST = {
+    "sphinxjp.themecore": "1.2"
+}  # type: Dict[unicode, unicode]
+
+
+class SphinxFactory(object):
```

Comment:
Indeed, it is not good naming.
question: do you mean splitting the feature to two classes?